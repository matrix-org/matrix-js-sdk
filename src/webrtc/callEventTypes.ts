// allow non-camelcase as these are events type that go onto the wire
/* eslint-disable camelcase */

import { CallErrorCode } from "./call";

// TODO: Change to "sdp_stream_metadata" when MSC3077 is merged
export const SDPStreamMetadataKey = "org.matrix.msc3077.sdp_stream_metadata";

export enum SDPStreamMetadataPurpose {
    Usermedia = "m.usermedia",
    Screenshare = "m.screenshare",
}

export interface SDPStreamMetadataObject {
    purpose: SDPStreamMetadataPurpose;
    audio_muted: boolean;
    video_muted: boolean;
}

export interface SDPStreamMetadata {
    [key: string]: SDPStreamMetadataObject;
}

export interface CallCapabilities {
    'm.call.transferee': boolean;
    'm.call.dtmf': boolean;
}

export interface CallReplacesTarget {
    id: string;
    display_name: string;
    avatar_url: string;
}

export interface MCallBase {
    call_id: string;
    version: string | number;
    party_id?: string;
    sender_session_id?: string;
    dest_session_id?: string;
}

export interface MCallAnswer extends MCallBase {
    answer: RTCSessionDescription;
    capabilities?: CallCapabilities;
    [SDPStreamMetadataKey]: SDPStreamMetadata;
}

export interface MCallSelectAnswer extends MCallBase {
    selected_party_id: string;
}

export interface MCallInviteNegotiate extends MCallBase {
    offer: RTCSessionDescription;
    description: RTCSessionDescription;
    lifetime: number;
    capabilities?: CallCapabilities;
    invitee?: string;
    sender_session_id?: string;
    dest_session_id?: string;
    [SDPStreamMetadataKey]: SDPStreamMetadata;
}

export interface MCallSDPStreamMetadataChanged extends MCallBase {
    [SDPStreamMetadataKey]: SDPStreamMetadata;
}

export interface MCallReplacesEvent extends MCallBase {
    replacement_id: string;
    target_user: CallReplacesTarget;
    create_call: string;
    await_call: string;
    target_room: string;
}

export interface MCAllAssertedIdentity extends MCallBase {
    asserted_identity: {
        id: string;
        display_name: string;
        avatar_url: string;
    };
}

export interface MCallCandidates extends MCallBase {
    candidates: RTCIceCandidate[];
}

export interface MCallHangupReject extends MCallBase {
    reason?: CallErrorCode;
}

/* eslint-enable camelcase */
