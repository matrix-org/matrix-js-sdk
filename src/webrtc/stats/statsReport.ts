/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { ConnectionStatsBandwidth, ConnectionStatsBitrate, PacketLoss } from "./connectionStats";
import { TransportStats } from "./transportStats";
import { Resolution } from "./media/mediaTrackStats";

export enum StatsReport {
    CONNECTION_STATS = "StatsReport.connection_stats",
    CALL_FEED_REPORT = "StatsReport.call_feed_report",
    BYTE_SENT_STATS = "StatsReport.byte_sent_stats",
    SUMMARY_STATS = "StatsReport.summary_stats",
}

/// ByteSentStatsReport ################################################################################################
export interface ByteSentStatsReport extends Map<TrackID, ByteSend> {
    callId?: string;
    opponentMemberId?: string;
    // is a map: `local trackID` => byte send
}

export type TrackID = string;
export type ByteSend = number;

/// ConnectionStatsReport ##############################################################################################
export interface ConnectionStatsReport {
    callId?: string;
    opponentMemberId?: string;
    bandwidth: ConnectionStatsBandwidth;
    bitrate: ConnectionStatsBitrate;
    packetLoss: PacketLoss;
    audioConcealment: Map<TrackID, AudioConcealment>;
    totalAudioConcealment: AudioConcealment;
    resolution: ResolutionMap;
    framerate: FramerateMap;
    codec: CodecMap;
    jitter: Map<TrackID, number>;
    transport: TransportStats[];
}

export interface AudioConcealment {
    concealedAudio: number;
    totalAudioDuration: number;
}

export interface ResolutionMap {
    local: Map<TrackID, Resolution>;
    remote: Map<TrackID, Resolution>;
}

export interface FramerateMap {
    local: Map<TrackID, number>;
    remote: Map<TrackID, number>;
}

export interface CodecMap {
    local: Map<TrackID, string>;
    remote: Map<TrackID, string>;
}

/// SummaryStatsReport #################################################################################################
export interface SummaryStatsReport {
    /**
     * Aggregated the information for percentage of received media
     *
     * This measure whether the current user receive data from a call participants.
     * As soon as a participant sends at least a byte media to this user, this counts as one measurement unit.
     * The units of measure divided by the total number of participants is a value between 0 and 1.
     */
    percentageReceivedMedia: number;
    percentageReceivedAudioMedia: number;
    percentageReceivedVideoMedia: number;
    maxJitter: number;
    maxPacketLoss: number;
    percentageConcealedAudio: number;
    peerConnections: number;
    opponentUsersInCall?: number;
    opponentDevicesInCall?: number;
    diffDevicesToPeerConnections?: number;
    ratioPeerConnectionToDevices?: number;
    // Todo: Decide if we want an index (or a timestamp) of this report in relation to the group call, to help differenciate when issues occur and ignore/track initial connection delays.
}

/// CallFeedReport #####################################################################################################
export interface CallFeedReport {
    callId: string;
    opponentMemberId: string;
    transceiver: TransceiverStats[];
    callFeeds: CallFeedStats[];
}

export interface CallFeedStats {
    stream: string;
    type: "remote" | "local";
    audio: TrackStats | null;
    video: TrackStats | null;
    purpose: string;
    prefix: string;
    isVideoMuted: boolean;
    isAudioMuted: boolean;
}

export interface TransceiverStats {
    readonly mid: string;
    readonly sender: TrackStats | null;
    readonly receiver: TrackStats | null;
    readonly direction: string;
    readonly currentDirection: string;
}

export interface TrackStats {
    readonly id: string;
    readonly kind: "audio" | "video";
    readonly settingDeviceId: string;
    readonly constrainDeviceId: string;
    readonly muted: boolean;
    readonly enabled: boolean;
    readonly readyState: "ended" | "live";
    readonly label: string;
}
