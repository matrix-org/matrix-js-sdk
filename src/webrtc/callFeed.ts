/*
Copyright 2021 Å imon Brandner <simon.bra.ag@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { SDPStreamMetadataPurpose } from "./callEventTypes";
import { acquireContext, releaseContext } from "./audioContext";
import { MatrixClient } from "../client";
import { RoomMember } from "../models/room-member";
import { logger } from "../logger";
import { TypedEventEmitter } from "../models/typed-event-emitter";
import { CallEvent, CallState, MatrixCall } from "./call";

const POLLING_INTERVAL = 200; // ms
export const SPEAKING_THRESHOLD = -60; // dB
const SPEAKING_SAMPLE_COUNT = 8; // samples

export interface ICallFeedOpts {
    client: MatrixClient;
    roomId?: string;
    userId: string;
    deviceId: string | undefined;
    stream: MediaStream;
    purpose: SDPStreamMetadataPurpose;
    /**
     * Whether or not the remote SDPStreamMetadata says audio is muted
     */
    audioMuted: boolean;
    /**
     * Whether or not the remote SDPStreamMetadata says video is muted
     */
    videoMuted: boolean;
    /**
     * The MatrixCall which is the source of this CallFeed
     */
    call?: MatrixCall;
}

export enum CallFeedEvent {
    NewStream = "new_stream",
    MuteStateChanged = "mute_state_changed",
    LocalVolumeChanged = "local_volume_changed",
    VolumeChanged = "volume_changed",
    ConnectedChanged = "connected_changed",
    Speaking = "speaking",
    Disposed = "disposed",
}

type EventHandlerMap = {
    [CallFeedEvent.NewStream]: (stream: MediaStream) => void;
    [CallFeedEvent.MuteStateChanged]: (audioMuted: boolean, videoMuted: boolean) => void;
    [CallFeedEvent.LocalVolumeChanged]: (localVolume: number) => void;
    [CallFeedEvent.VolumeChanged]: (volume: number) => void;
    [CallFeedEvent.ConnectedChanged]: (connected: boolean) => void;
    [CallFeedEvent.Speaking]: (speaking: boolean) => void;
    [CallFeedEvent.Disposed]: () => void;
};

export class CallFeed extends TypedEventEmitter<CallFeedEvent, EventHandlerMap> {
    public stream: MediaStream;
    public sdpMetadataStreamId: string;
    public userId: string;
    public readonly deviceId: string | undefined;
    public purpose: SDPStreamMetadataPurpose;
    public speakingVolumeSamples: number[];

    private client: MatrixClient;
    private call?: MatrixCall;
    private roomId?: string;
    private audioMuted: boolean;
    private videoMuted: boolean;
    private localVolume = 1;
    private measuringVolumeActivity = false;
    private audioContext?: AudioContext;
    private analyser?: AnalyserNode;
    private frequencyBinCount?: Float32Array;
    private speakingThreshold = SPEAKING_THRESHOLD;
    private speaking = false;
    private volumeLooperTimeout?: ReturnType<typeof setTimeout>;
    private _disposed = false;
    private _connected = false;

    public constructor(opts: ICallFeedOpts) {
        super();

        this.client = opts.client;
        this.call = opts.call;
        this.roomId = opts.roomId;
        this.userId = opts.userId;
        this.deviceId = opts.deviceId;
        this.purpose = opts.purpose;
        this.audioMuted = opts.audioMuted;
        this.videoMuted = opts.videoMuted;
        this.speakingVolumeSamples = new Array(SPEAKING_SAMPLE_COUNT).fill(-Infinity);
        this.sdpMetadataStreamId = opts.stream.id;

        this.updateStream(null, opts.stream);
        this.stream = opts.stream; // updateStream does this, but this makes TS happier

        if (this.hasAudioTrack) {
            this.initVolumeMeasuring();
        }

        if (opts.call) {
            opts.call.addListener(CallEvent.State, this.onCallState);
            this.onCallState(opts.call.state);
        }
    }

    public get connected(): boolean {
        // Local feeds are always considered connected
        return this.isLocal() || this._connected;
    }

    private set connected(connected: boolean) {
        this._connected = connected;
        this.emit(CallFeedEvent.ConnectedChanged, this.connected);
    }

    public get hasAudioTrack(): boolean {
        return this.stream.getAudioTracks().length > 0;
    }

    private updateStream(oldStream: MediaStream | null, newStream: MediaStream): void {
        if (newStream === oldStream) return;

        const wasMeasuringVolumeActivity = this.measuringVolumeActivity;

        if (oldStream) {
            oldStream.removeEventListener("addtrack", this.onAddTrack);
            this.measureVolumeActivity(false);
        }

        this.stream = newStream;
        newStream.addEventListener("addtrack", this.onAddTrack);

        if (this.hasAudioTrack) {
            this.initVolumeMeasuring();
            if (wasMeasuringVolumeActivity) this.measureVolumeActivity(true);
        } else {
            this.measureVolumeActivity(false);
        }

        this.emit(CallFeedEvent.NewStream, this.stream);
    }

    private initVolumeMeasuring(): void {
        if (!this.hasAudioTrack) return;
        if (!this.audioContext) this.audioContext = acquireContext();

        this.analyser = this.audioContext.createAnalyser();
        this.analyser.fftSize = 512;
        this.analyser.smoothingTimeConstant = 0.1;

        const mediaStreamAudioSourceNode = this.audioContext.createMediaStreamSource(this.stream);
        mediaStreamAudioSourceNode.connect(this.analyser);

        this.frequencyBinCount = new Float32Array(this.analyser.frequencyBinCount);
    }

    private onAddTrack = (): void => {
        this.emit(CallFeedEvent.NewStream, this.stream);
    };

    private onCallState = (state: CallState): void => {
        if (state === CallState.Connected) {
            this.connected = true;
        } else if (state === CallState.Connecting) {
            this.connected = false;
        }
    };

    /**
     * Returns callRoom member
     * @returns member of the callRoom
     */
    public getMember(): RoomMember | null {
        const callRoom = this.client.getRoom(this.roomId);
        return callRoom?.getMember(this.userId) ?? null;
    }

    /**
     * Returns true if CallFeed is local, otherwise returns false
     * @returns is local?
     */
    public isLocal(): boolean {
        return (
            this.userId === this.client.getUserId() &&
            (this.deviceId === undefined || this.deviceId === this.client.getDeviceId())
        );
    }

    /**
     * Returns true if audio is muted or if there are no audio
     * tracks, otherwise returns false
     * @returns is audio muted?
     */
    public isAudioMuted(): boolean {
        return this.stream.getAudioTracks().length === 0 || this.audioMuted;
    }

    /**
     * Returns true video is muted or if there are no video
     * tracks, otherwise returns false
     * @returns is video muted?
     */
    public isVideoMuted(): boolean {
        // We assume only one video track
        return this.stream.getVideoTracks().length === 0 || this.videoMuted;
    }

    public isSpeaking(): boolean {
        return this.speaking;
    }

    /**
     * Replaces the current MediaStream with a new one.
     * The stream will be different and new stream as remote parties are
     * concerned, but this can be used for convenience locally to set up
     * volume listeners automatically on the new stream etc.
     * @param newStream - new stream with which to replace the current one
     */
    public setNewStream(newStream: MediaStream): void {
        this.updateStream(this.stream, newStream);
    }

    /**
     * Set one or both of feed's internal audio and video video mute state
     * Either value may be null to leave it as-is
     * @param audioMuted - is the feed's audio muted?
     * @param videoMuted - is the feed's video muted?
     */
    public setAudioVideoMuted(audioMuted: boolean | null, videoMuted: boolean | null): void {
        if (audioMuted !== null) {
            if (this.audioMuted !== audioMuted) {
                this.speakingVolumeSamples.fill(-Infinity);
            }
            this.audioMuted = audioMuted;
        }
        if (videoMuted !== null) this.videoMuted = videoMuted;
        this.emit(CallFeedEvent.MuteStateChanged, this.audioMuted, this.videoMuted);
    }

    /**
     * Starts emitting volume_changed events where the emitter value is in decibels
     * @param enabled - emit volume changes
     */
    public measureVolumeActivity(enabled: boolean): void {
        if (enabled) {
            if (!this.analyser || !this.frequencyBinCount || !this.hasAudioTrack) return;

            this.measuringVolumeActivity = true;
            this.volumeLooper();
        } else {
            this.measuringVolumeActivity = false;
            this.speakingVolumeSamples.fill(-Infinity);
            this.emit(CallFeedEvent.VolumeChanged, -Infinity);
        }
    }

    public setSpeakingThreshold(threshold: number): void {
        this.speakingThreshold = threshold;
    }

    private volumeLooper = (): void => {
        if (!this.analyser) return;

        if (!this.measuringVolumeActivity) return;

        this.analyser.getFloatFrequencyData(this.frequencyBinCount!);

        let maxVolume = -Infinity;
        for (const volume of this.frequencyBinCount!) {
            if (volume > maxVolume) {
                maxVolume = volume;
            }
        }

        this.speakingVolumeSamples.shift();
        this.speakingVolumeSamples.push(maxVolume);

        this.emit(CallFeedEvent.VolumeChanged, maxVolume);

        let newSpeaking = false;

        for (const volume of this.speakingVolumeSamples) {
            if (volume > this.speakingThreshold) {
                newSpeaking = true;
                break;
            }
        }

        if (this.speaking !== newSpeaking) {
            this.speaking = newSpeaking;
            this.emit(CallFeedEvent.Speaking, this.speaking);
        }

        this.volumeLooperTimeout = setTimeout(this.volumeLooper, POLLING_INTERVAL);
    };

    public clone(): CallFeed {
        const mediaHandler = this.client.getMediaHandler();
        const stream = this.stream.clone();
        logger.log(`CallFeed clone() cloning stream (originalStreamId=${this.stream.id}, newStreamId${stream.id})`);

        if (this.purpose === SDPStreamMetadataPurpose.Usermedia) {
            mediaHandler.userMediaStreams.push(stream);
        } else {
            mediaHandler.screensharingStreams.push(stream);
        }

        return new CallFeed({
            client: this.client,
            roomId: this.roomId,
            userId: this.userId,
            deviceId: this.deviceId,
            stream,
            purpose: this.purpose,
            audioMuted: this.audioMuted,
            videoMuted: this.videoMuted,
        });
    }

    public dispose(): void {
        clearTimeout(this.volumeLooperTimeout);
        this.stream?.removeEventListener("addtrack", this.onAddTrack);
        this.call?.removeListener(CallEvent.State, this.onCallState);
        if (this.audioContext) {
            this.audioContext = undefined;
            this.analyser = undefined;
            releaseContext();
        }
        this._disposed = true;
        this.emit(CallFeedEvent.Disposed);
    }

    public get disposed(): boolean {
        return this._disposed;
    }

    private set disposed(value: boolean) {
        this._disposed = value;
    }

    public getLocalVolume(): number {
        return this.localVolume;
    }

    public setLocalVolume(localVolume: number): void {
        this.localVolume = localVolume;
        this.emit(CallFeedEvent.LocalVolumeChanged, localVolume);
    }
}
