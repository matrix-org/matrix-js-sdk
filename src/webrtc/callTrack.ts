/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { randomString } from "../randomstring";
import { SDPStreamMetadataTrack } from "./callEventTypes";

export interface CallTrackOpts {}

export abstract class CallTrack {
    public abstract get id(): string | undefined;
    public abstract get track(): MediaStreamTrack | undefined;
    public abstract get trackId(): string | undefined;
    public abstract get metadata(): SDPStreamMetadataTrack | undefined;
    public abstract get kind(): string | undefined;
    public abstract get muted(): boolean;

    protected readonly _id: string;
    protected _transceiver?: RTCRtpTransceiver;

    public constructor(opts: CallTrackOpts) {
        this._id = randomString(32);
    }

    public get transceiver(): RTCRtpTransceiver | undefined {
        return this._transceiver;
    }

    public get isAudio(): boolean {
        return this.kind === "audio";
    }

    public get isVideo(): boolean {
        return this.kind === "video";
    }
}
