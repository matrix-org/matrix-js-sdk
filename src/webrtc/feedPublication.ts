/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixCall } from "./call";
import { SDPStreamMetadataObject, SDPStreamMetadataTracks } from "./callEventTypes";
import { LocalCallFeed } from "./localCallFeed";
import { TrackPublication } from "./trackPublication";

interface CallFeedPublicationOpts {
    call: MatrixCall;
    feed: LocalCallFeed;
}

/**
 * FeedPublication represents a LocalCallFeed being published to a specific peer
 * connection. It stores an array of track publications. This class needs to
 * exist, so that we are able elegantly retrieve feed's track publications on a
 * given peer connection.
 */
export class FeedPublication {
    public readonly call: MatrixCall;
    public readonly feed: LocalCallFeed;
    private trackPublications: TrackPublication[] = [];

    public constructor(opts: CallFeedPublicationOpts) {
        this.call = opts.call;
        this.feed = opts.feed;
    }

    public get metadata(): SDPStreamMetadataObject {
        return {
            user_id: this.feed.userId,
            device_id: this.feed.deviceId,
            purpose: this.feed.purpose,
            audio_muted: this.feed.audioMuted,
            video_muted: this.feed.videoMuted,
            tracks: this.trackPublications.reduce(
                (metadata: SDPStreamMetadataTracks, publication: TrackPublication) => {
                    if (!publication.trackId) return metadata;

                    metadata[publication.trackId] = publication.metadata;
                    return metadata;
                },
                {},
            ),
        };
    }

    public get streamId(): string | undefined {
        return this.trackPublications[0]?.streamId;
    }

    public addTrackPublication(publication: TrackPublication): void {
        this.trackPublications.push(publication);
    }

    public removeTrackPublication(publication: TrackPublication): void {
        this.trackPublications.splice(this.trackPublications.indexOf(publication), 1);
    }
}
