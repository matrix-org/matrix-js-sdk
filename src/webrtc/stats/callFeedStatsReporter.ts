/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { CallFeedReport, CallFeedStats, TrackStats, TransceiverStats } from "./statsReport";
import { CallFeed } from "../callFeed";

export class CallFeedStatsReporter {
    public static buildCallFeedReport(callId: string, opponentMemberId: string, pc: RTCPeerConnection): CallFeedReport {
        const rtpTransceivers = pc.getTransceivers();
        const transceiver: TransceiverStats[] = [];
        const callFeeds: CallFeedStats[] = [];

        rtpTransceivers.forEach((t) => {
            const sender = t.sender?.track ? CallFeedStatsReporter.buildTrackStats(t.sender.track, "sender") : null;
            const receiver = CallFeedStatsReporter.buildTrackStats(t.receiver.track, "receiver");
            transceiver.push({
                mid: t.mid == null ? "null" : t.mid,
                direction: t.direction,
                currentDirection: t.currentDirection == null ? "null" : t.currentDirection,
                sender,
                receiver,
            });
        });

        return {
            callId,
            opponentMemberId,
            transceiver,
            callFeeds,
        };
    }

    private static buildTrackStats(track: MediaStreamTrack, label = "--"): TrackStats {
        const settingDeviceId = track.getSettings()?.deviceId;
        const constrainDeviceId = track.getConstraints()?.deviceId;

        return {
            id: track.id,
            kind: track.kind,
            settingDeviceId: settingDeviceId ? settingDeviceId : "unknown",
            constrainDeviceId: constrainDeviceId ? constrainDeviceId : "unknown",
            muted: track.muted,
            enabled: track.enabled,
            readyState: track.readyState,
            label,
        } as TrackStats;
    }

    public static expandCallFeedReport(
        report: CallFeedReport,
        callFeeds: CallFeed[],
        prefix = "unknown",
    ): CallFeedReport {
        if (!report.callFeeds) {
            report.callFeeds = [];
        }
        callFeeds.forEach((feed) => {
            const audioTracks = feed.stream.getAudioTracks();
            const videoTracks = feed.stream.getVideoTracks();
            const audio =
                audioTracks.length > 0
                    ? CallFeedStatsReporter.buildTrackStats(feed.stream.getAudioTracks()[0], feed.purpose)
                    : null;
            const video =
                videoTracks.length > 0
                    ? CallFeedStatsReporter.buildTrackStats(feed.stream.getVideoTracks()[0], feed.purpose)
                    : null;
            const feedStats = {
                stream: feed.stream.id,
                type: feed.isLocal() ? "local" : "remote",
                audio,
                video,
                purpose: feed.purpose,
                prefix,
                isVideoMuted: feed.isVideoMuted(),
                isAudioMuted: feed.isAudioMuted(),
            } as CallFeedStats;
            report.callFeeds.push(feedStats);
        });
        return report;
    }
}
