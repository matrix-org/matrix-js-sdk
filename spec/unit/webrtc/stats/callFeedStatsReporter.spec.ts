/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { CallFeedStatsReporter } from "../../../../src/webrtc/stats/callFeedStatsReporter";
import { CallFeedReport } from "../../../../src/webrtc/stats/statsReport";
import { CallFeed } from "../../../../src/webrtc/callFeed";

const CALL_ID = "CALL_ID";
const USER_ID = "USER_ID";
describe("CallFeedStatsReporter", () => {
    let rtcSpy: RTCPeerConnection;
    beforeEach(() => {
        rtcSpy = {} as RTCPeerConnection;
        rtcSpy.getTransceivers = jest.fn().mockReturnValue(buildTransceiverMocks());
    });

    describe("should", () => {
        it("builds CallFeedReport", async () => {
            expect(CallFeedStatsReporter.buildCallFeedReport(CALL_ID, USER_ID, rtcSpy)).toMatchSnapshot();
        });

        it("extends CallFeedReport with call feeds", async () => {
            const feed = buildCallFeedMock("1");
            const callFeedList: CallFeed[] = [feed];
            const report = {
                callId: "callId",
                opponentMemberId: "opponentMemberId",
                transceiver: [],
                callFeeds: [],
            } as CallFeedReport;

            expect(CallFeedStatsReporter.expandCallFeedReport(report, callFeedList).callFeeds).toMatchSnapshot();
        });
    });

    const buildTransceiverMocks = (): RTCRtpTransceiver[] => {
        const trans1 = {
            mid: "0",
            direction: "sendrecv",
            currentDirection: "sendonly",
            sender: buildSenderMock("sender_audio_0", "audio"),
            receiver: buildReceiverMock("receiver_audio_0", "audio"),
        } as RTCRtpTransceiver;
        const trans2 = {
            mid: "1",
            direction: "recvonly",
            currentDirection: "sendrecv",
            sender: buildSenderMock("sender_video_1", "video"),
            receiver: buildReceiverMock("receiver_video_1", "video"),
        } as RTCRtpTransceiver;
        const trans3 = {
            mid: "2",
            direction: "recvonly",
            currentDirection: "recvonly",
            sender: { track: null } as RTCRtpSender,
            receiver: buildReceiverMock("receiver_video_2", "video"),
        } as RTCRtpTransceiver;
        return [trans1, trans2, trans3];
    };

    const buildSenderMock = (id: string, kind: "audio" | "video"): RTCRtpSender => {
        const track = buildTrackMock(id, kind);
        return {
            track,
        } as RTCRtpSender;
    };

    const buildReceiverMock = (id: string, kind: "audio" | "video"): RTCRtpReceiver => {
        const track = buildTrackMock(id, kind);
        return {
            track,
        } as RTCRtpReceiver;
    };

    const buildTrackMock = (id: string, kind: "audio" | "video"): MediaStreamTrack => {
        return {
            id,
            kind,
            enabled: true,
            label: "--",
            muted: false,
            readyState: "live",
            getSettings: () => ({ deviceId: `settingDeviceId-${id}` }),
            getConstraints: () => ({ deviceId: `constrainDeviceId-${id}` }),
        } as MediaStreamTrack;
    };

    const buildCallFeedMock = (id: string, isLocal = true): CallFeed => {
        const stream = {
            id: `stream-${id}`,
            getAudioTracks(): MediaStreamTrack[] {
                return [buildTrackMock(`video-${id}`, "video")];
            },
            getVideoTracks(): MediaStreamTrack[] {
                return [buildTrackMock(`audio-${id}`, "audio")];
            },
        } as MediaStream;
        return {
            stream,
            isLocal: () => isLocal,
            isVideoMuted: () => false,
            isAudioMuted: () => true,
        } as CallFeed;
    };
});
