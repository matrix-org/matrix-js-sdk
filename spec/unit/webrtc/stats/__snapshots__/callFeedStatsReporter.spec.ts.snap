// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallFeedStatsReporter should builds CallFeedReport 1`] = `
{
  "callFeeds": [],
  "callId": "CALL_ID",
  "opponentMemberId": "USER_ID",
  "transceiver": [
    {
      "currentDirection": "sendonly",
      "direction": "sendrecv",
      "mid": "0",
      "receiver": {
        "constrainDeviceId": "constrainDeviceId-receiver_audio_0",
        "enabled": true,
        "id": "receiver_audio_0",
        "kind": "audio",
        "label": "receiver",
        "muted": false,
        "readyState": "live",
        "settingDeviceId": "settingDeviceId-receiver_audio_0",
      },
      "sender": {
        "constrainDeviceId": "constrainDeviceId-sender_audio_0",
        "enabled": true,
        "id": "sender_audio_0",
        "kind": "audio",
        "label": "sender",
        "muted": false,
        "readyState": "live",
        "settingDeviceId": "settingDeviceId-sender_audio_0",
      },
    },
    {
      "currentDirection": "sendrecv",
      "direction": "recvonly",
      "mid": "1",
      "receiver": {
        "constrainDeviceId": "constrainDeviceId-receiver_video_1",
        "enabled": true,
        "id": "receiver_video_1",
        "kind": "video",
        "label": "receiver",
        "muted": false,
        "readyState": "live",
        "settingDeviceId": "settingDeviceId-receiver_video_1",
      },
      "sender": {
        "constrainDeviceId": "constrainDeviceId-sender_video_1",
        "enabled": true,
        "id": "sender_video_1",
        "kind": "video",
        "label": "sender",
        "muted": false,
        "readyState": "live",
        "settingDeviceId": "settingDeviceId-sender_video_1",
      },
    },
    {
      "currentDirection": "recvonly",
      "direction": "recvonly",
      "mid": "2",
      "receiver": {
        "constrainDeviceId": "constrainDeviceId-receiver_video_2",
        "enabled": true,
        "id": "receiver_video_2",
        "kind": "video",
        "label": "receiver",
        "muted": false,
        "readyState": "live",
        "settingDeviceId": "settingDeviceId-receiver_video_2",
      },
      "sender": null,
    },
  ],
}
`;

exports[`CallFeedStatsReporter should extends CallFeedReport with call feeds 1`] = `
[
  {
    "audio": {
      "constrainDeviceId": "constrainDeviceId-video-1",
      "enabled": true,
      "id": "video-1",
      "kind": "video",
      "label": "--",
      "muted": false,
      "readyState": "live",
      "settingDeviceId": "settingDeviceId-video-1",
    },
    "isAudioMuted": true,
    "isVideoMuted": false,
    "prefix": "unknown",
    "purpose": undefined,
    "stream": "stream-1",
    "type": "local",
    "video": {
      "constrainDeviceId": "constrainDeviceId-audio-1",
      "enabled": true,
      "id": "audio-1",
      "kind": "audio",
      "label": "--",
      "muted": false,
      "readyState": "live",
      "settingDeviceId": "settingDeviceId-audio-1",
    },
  },
]
`;
